File: proposed-projects/round-trip-experiment/reverse/20260909T023701Z_e94dca40/20260910T060603Z_391335fe/ported_implementation/gbnf/rules_builder/symbol_ids.py
from __future__ import annotations

from typing import Iterator


class SymbolIds:
    def __init__(self) -> None:
        self._map: dict[str, int] = {}
        self._positions: dict[str, int] = {}
        self._reverse_map: dict[int, str] = {}

    def __iter__(self) -> Iterator[tuple[str, int]]:
        yield from self._map.items()

    def entries(self) -> Iterator[tuple[str, int]]:
        return iter(self._map.items())

    @property
    def size(self) -> int:
        return len(self._map)

    def get(self, key: str) -> int | None:
        return self._map.get(key)

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._reverse_map[value] = key
        self._positions[key] = pos

    def has(self, key: str) -> bool:
        return key in self._map

    def reverse_get(self, key: int) -> str:
        val = self._reverse_map.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain value: {key}")
        return val

    def get_pos(self, key: str) -> int:
        val = self._positions.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain key: {key}")
        return val

File: packages/gbnf-experiment/data/20260909T123239Z_5982b634/ported_implementation/src/rules-builder/symbol-ids.ts
export class SymbolIds {
  #map = new Map<string, number>();
  #pos = new Map<string, number>();
  #reverseMap = new Map<number, string>();

  get size(): number {
    return this.#map.size;
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  entries(): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  has(key: string): boolean {
    return this.#map.has(key);
  }

  get(key: string): number | undefined {
    return this.#map.get(key);
  }

  set(key: string, value: number, pos: number): void {
    this.#map.set(key, value);
    this.#reverseMap.set(value, key);
    this.#pos.set(key, pos);
  }

  reverseGet(key: number): string {
    const val = this.#reverseMap.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return val;
  }

  getPos(key: string): number {
    const val = this.#pos.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return val;
  }
}
